pool/pool.go

```diff
@@ -30,284 +31,576 @@ import (
+// as it leads to undefined behavior.
+//
+// Each method which produces a NeoFS API call may return an error.
+// Status of underlying server response is casted to built-in error instance.
+// Certain statuses can be checked using `sdkClient` and standard `errors` packages.
+// Note that package provides some helper functions to work with status returns
+// (e.g. sdkClient.IsErrContainerNotFound, sdkClient.IsErrObjectNotFound).
+//
+// See pool package overview to get some examples.
 type Pool struct {
 	innerPools      []*innerPool
 	key             *ecdsa.PrivateKey
 	owner           *owner.ID
 	cancel          context.CancelFunc
 	closedCh        chan struct{}
-	cache           *SessionCache
+	cache           *sessionCache
 	stokenDuration  uint64
 	stokenThreshold time.Duration
+	rebalanceParams rebalanceParameters
+	clientBuilder   func(endpoint string) (client, error)
+	logger          *zap.Logger
 }
 
 type innerPool struct {
 	lock        sync.RWMutex
-	sampler     *Sampler
+	sampler     *sampler
 	clientPacks []*clientPack
 }
 
 const (
 	defaultSessionTokenExpirationDuration = 100 // in blocks
 
 	defaultSessionTokenThreshold = 5 * time.Second
+	defaultRebalanceInterval     = 25 * time.Second
+	defaultRequestTimeout        = 4 * time.Second
 )
 
-func newPool(ctx context.Context, options *BuilderOptions) (*Pool, error) {
-	cache, err := NewCache()
+// NewPool create connection pool using parameters.
+func NewPool(options InitParameters) (*Pool, error) {
+	if options.key == nil {
+		return nil, fmt.Errorf("missed required parameter 'Key'")
+	}
+
+	nodesParams, err := adjustNodeParams(options.nodeParams)
 	if err != nil {
-		return nil, fmt.Errorf("couldn't create cache: %w", err)
+		return nil, err
 	}
 
-	if options.SessionExpirationDuration == 0 {
-		options.SessionExpirationDuration = defaultSessionTokenExpirationDuration
+	fillDefaultInitParams(&options)
+
+	cache, err := newCache()
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create cache: %w", err)
 	}
 
-	if options.SessionTokenThreshold <= 0 {
-		options.SessionTokenThreshold = defaultSessionTokenThreshold
+	pool := &Pool{
+		key:             options.key,
+		owner:           owner.NewIDFromPublicKey(&options.key.PublicKey),
+		cache:           cache,
+		logger:          options.logger,
+		stokenDuration:  options.sessionExpirationDuration,
+		stokenThreshold: options.sessionTokenThreshold,
+		rebalanceParams: rebalanceParameters{
+			nodesParams:               nodesParams,
+			nodeRequestTimeout:        options.nodeRequestTimeout,
+			clientRebalanceInterval:   options.clientRebalanceInterval,
+			sessionExpirationDuration: options.sessionExpirationDuration,
+		},
+		clientBuilder: options.clientBuilder,
 	}
 
-	ownerID := owner.NewIDFromPublicKey(&options.Key.PublicKey)
+	return pool, nil
+}
 
-	inner := make([]*innerPool, len(options.nodesParams))
+// Dial establishes a connection to the servers from the NeoFS network.
```

Comment:
I'd also mention that it starts rebalancing routine, and add a mutual reference to the method
```
// SetClientRebalanceInterval ...
//
// See also Pool.Dial.
func (x *InitParameters) SetClientRebalanceInterval(interval time.Duration)
```
```
// Dial ...
//
// See also InitParameters.SetClientRebalanceInterval.
func (p *Pool) Dial(ctx context.Context) error
```